### PR TITLE
Feature/refactor

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR fixes some docstrings and public access patterns.

## Fixes

-

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/helpers/dialogue/base.py [hunk 1]
```diff
@@ -434,18 +433,30 @@ def _additional_rules(self, message: Message) -> bool:
             result = target == last_target + 1
         return result
 
+    def update_dialogue_label(self, final_dialogue_label: DialogueLabel) -> None:
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:436]: Write some tests for this

File: aea/helpers/dialogue/base.py [hunk 2]
```diff
@@ -495,27 +502,6 @@ def __str__(self) -> str:
         representation = representation[:-1]
         return representation
 
-    # ToDo the following methods are left for backwards compatibility reasons and are unsafe to use. They will be removed in the future
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:498]: Removed as no longer required

File: aea/helpers/dialogue/base.py [hunk 3]
```diff
@@ -741,76 +725,3 @@ def _next_dialogue_nonce(self) -> int:
         """
         self._dialogue_nonce += 1
         return self._dialogue_nonce
-
-    # TODO the following methods are left for backwards compatibility reasons and will be removed in the future
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:745]: No longer required

File: aea/protocols/base.py [hunk 4]
```diff
@@ -114,12 +115,30 @@ def body(self, body: Dict) -> None:
...
+        """Get the performative of the message."""
+        assert self.is_set("performative"), "performative is not set."
+        return cast(Enum, self.get("performative"))
+
+    @property
+    def target(self) -> int:
+        """Get the target of the message."""
+        assert self.is_set("target"), "target is not set."
+        return cast(int, self.get("target"))
+
```
Comment (DavidMinarsch) [inline aea/protocols/base.py:141]: The base dialogue requires these properties. There is no harm of setting them in the base class as we expect them to be present in all messages anyway as per current generator.

File: aea/aea_builder.py [hunk 5]
```diff
@@ -589,7 +589,7 @@ def add_component(
         self._check_can_add(configuration)
         # update dependency graph
         self._package_dependency_manager.add_component(configuration)
-        configuration._directory = directory
+        configuration._directory = directory  # pylint: disable=protected-access
```
Comment (marcofavorito) [inline aea/aea_builder.py:592]: Optional, minor comment: what about a one-time setter?

```python
@directory.setter
def directory(self, directory: str) -> None:
    assert self._directory is None, "Directory already set"
    self._directory = directory
```

Comment (DavidMinarsch) [reply aea/aea_builder.py:592]: Sounds good!

File: aea/helpers/async_friendly_queue.py [hunk 6]
```diff
@@ -41,7 +41,9 @@ def put(self, item: Any, *args, **kwargs) -> None:
         super().put(item, *args, **kwargs)
         if self._non_empty_waiters:
             waiter = self._non_empty_waiters.popleft()
-            waiter._loop.call_soon_threadsafe(waiter.set_result, True)
+            waiter._loop.call_soon_threadsafe(  # pylint: disable=protected-access
```
Comment (marcofavorito) [inline aea/helpers/async_friendly_queue.py:44]: If we didn't have to be compatible with Python 3.6, we could have used `get_loop()`:
https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.get_loop

File: aea/decision_maker/default.py [hunk 7]
```diff
@@ -246,7 +246,7 @@ def apply_transactions(
         """
         new_state = copy.copy(self)
         for tx_message in transactions:
-            new_state._update(tx_message)
+            new_state._update(tx_message)  # pylint: disable=protected-access
```
Comment (marcofavorito) [inline aea/decision_maker/default.py:249]: does `_update` has to be private? O/w make it public

File: aea/helpers/async_utils.py [hunk 8]
```diff
@@ -127,7 +127,7 @@ async def wait(self, state_or_states: Union[Any, Sequence[Any]]) -> Tuple[Any, A
             return (None, self._state)
 
         watcher: Future = Future()
-        watcher._states = states  # type: ignore
+        watcher._states = states  # type: ignore  # pylint: disable=protected-access
```
Comment (marcofavorito) [inline aea/helpers/async_utils.py:130]: I don't think there are counterarguments in making the `_states` attribute to be public. It is a custom attribute, so it won't be used by the `asyncio` library (I would be very surprised if it is).

Moreover, an ideal approach would be to extend the `Future` class to include such custom attributes.

Comment (marcofavorito) [reply aea/helpers/async_utils.py:130]: This, of course, will affect the above code, which must be changed accordingly. But that would save some Pylint `disable `directives.

File: aea/helpers/dialogue/base.py [hunk 9]
```diff
@@ -391,20 +386,20 @@ def _basic_rules(self, message: Message) -> bool:
         :return: True if valid, False otherwise.
         """
         message_id = message.message_id
-        target = cast(int, message.get("target"))
-        performative = cast(Enum, message.get("performative"))
+        target = message.target
+        performative = message.performative
 
-        if self.is_empty:
+        if self.last_message is None:
```
Comment (marcofavorito) [inline aea/helpers/dialogue/base.py:392]: why did you change this?

Comment (DavidMinarsch) [reply aea/helpers/dialogue/base.py:392]: Because the implementation meant one had to then check again if the last message was present for pypi. It makes more sense this way.

File: aea/helpers/async_utils.py (hunk 8 above)
Comment (DavidMinarsch) [reply aea/helpers/async_utils.py:130]: Good points; although I'd rather not address them now as I don't want to open a can of worms.

File: aea/decision_maker/default.py (hunk 7 above)
Comment (DavidMinarsch) [reply aea/decision_maker/default.py:249]: It can be public. Will fix.

File: aea/helpers/async_friendly_queue.py (hunk 6 above)
Comment (DavidMinarsch) [reply aea/helpers/async_friendly_queue.py:44]: But we do have to be compatible :)

File: aea/protocols/base.py (hunk 4 above)
Comment (marcofavorito) [reply aea/protocols/base.py:141]: OK, so why in the future we don't make the class `Message` less generic? That is, the constructor requires the `performative`, `target` etc.?

Sometimes for a framework, it is good to "take decisions" and force the user to think in the framework's terms.

File: aea/helpers/search/models.py [hunk 10]
```diff
@@ -605,7 +609,7 @@ def _check_validity(self):
                 "subexpression must be at least 2.".format(type(self).__name__)
             )
         for constraint in self.constraints:
-            constraint._check_validity()
+            constraint._check_validity()  # pylint: disable=protected-access
```
Comment (marcofavorito) [inline aea/helpers/search/models.py:612]: maybe also this one could be made public. As far as I can see, there are no side effects in the implementation of that method.

File: aea/protocols/base.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/protocols/base.py:141]: Yes that can be done; although it's probably best to postpone this until later when we have more certainty on the required fields.

File: aea/helpers/search/models.py (hunk 10 above)
Comment (DavidMinarsch) [reply aea/helpers/search/models.py:612]: Yeah we can do

File: benchmark/framework/aea_test_wrapper.py [hunk 11]
```diff
@@ -168,15 +167,15 @@ def set_loop_timeout(self, timeout: float) -> None:
 
         :return: None
         """
-        self.aea._timeout = timeout
+        self.aea._timeout = timeout  # pylint: disable=protected-access
```
Comment (marcofavorito) [inline benchmark/framework/aea_test_wrapper.py:170]: setter? Although we shouldn't allow undesired change of timeout... So probably better leave it for now.

File: tests/test_configurations/test_aea_config.py [hunk 12]
```diff
@@ -164,7 +164,7 @@ def test_builder_applies_default_value_to_aea(self, mock: Mock) -> None:
         """Test AEABuilder applies default value to AEA instance when option is not specified in config."""
         configuration = self._make_configuration(NotSet)
         builder = AEABuilder()
-        builder._set_from_configuration(configuration, aea_project_path=Path("."))
+        builder.set_from_configuration(configuration, aea_project_path=Path("."))
```
Comment (marcofavorito) [inline tests/test_configurations/test_aea_config.py:167]: If public, this might lead to undesired side-effects. 
The implementation "adds" stuff (components/private-keys etc) to the builder, doesn't replace it.

Comment (marcofavorito) [reply tests/test_configurations/test_aea_config.py:167]: An easy solution is to make it private again.

File: benchmark/framework/aea_test_wrapper.py (hunk 11 above)
Comment (DavidMinarsch) [reply benchmark/framework/aea_test_wrapper.py:170]: Yes I think it should be left as is (this is a unique exception imo)
